net/respondd/src/respondd.c

```diff
@@ -41,150 +48,381 @@
+	if (setsockopt(sock, IPPROTO_IPV6, IPV6_JOIN_GROUP, &mreq, sizeof(mreq)) == -1)
+		goto error;
 
-  return;
+	return;
 
  error:
-  fprintf(stderr, "Could not join multicast group on %s: ", iface);
-  perror(NULL);
+	fprintf(stderr, "Could not join multicast group on %s: ", iface);
+	perror(NULL);
+}
+
+
+static void update_time(void) {
+	struct timespec tp;
+	clock_gettime(CLOCK_MONOTONIC, &tp);
+
+	now = (int64_t)tp.tv_sec * 1000 + tp.tv_nsec / 1000000;
+}
+
+
+/** Merges two JSON objects (on conflicts, object a will be preferred) */
+static struct json_object * merge_json(struct json_object *a, struct json_object *b) {
+	if (!json_object_is_type(a, json_type_object) || !json_object_is_type(b, json_type_object)) {
+		json_object_put(b);
+		return a;
+	}
+
+	json_object_object_foreach(a, key, val_a) {
+		struct json_object *val_b;
+
+		json_object_get(val_a);
+
+		if (!json_object_object_get_ex(b, key, &val_b)) {
+			json_object_object_add(b, key, val_a);
+			continue;
+		}
+
+		json_object_get(val_b);
+
+		json_object_object_add(b, key, merge_json(val_a, val_b));
+	}
+
+	json_object_put(a);
+	return b;
+}
+
+static const struct respondd_provider_info * get_providers(const char *filename) {
+	char path[2 + strlen(filename) + 1];
+	snprintf(path, sizeof(path), "./%s", filename);
+
+	void *handle = dlopen(path, RTLD_NOW|RTLD_LOCAL);
+	if (!handle)
+		return NULL;
+
+	const struct respondd_provider_info *ret = dlsym(handle, "respondd_providers");
+	if (!ret) {
+		dlclose(handle);
+		return NULL;
+	}
+
+	return ret;
+}
+
+static void load_cache_time(struct request_type *r, const char *name) {
+	r->cache_time = 0;
+
+	char filename[strlen(name) + 7];
+	snprintf(filename, sizeof(filename), "%s.cache", name);
+
+	FILE *f = fopen(filename, "r");
+	if (!f)
+		return;
+
+	float val = 0;
+	fscanf(f, "%f", &val);
+	fclose(f);
+
+	if (val > 0) {
+		r->cache = NULL;
+		r->cache_time = (uint64_t)(val * 1000);
+		r->cache_timeout = now;
+	}
 }
 
-static void serve(lua_State *L, int sock) {
-  char buffer[256];
-  ssize_t read_bytes;
-  struct timespec tp;
+static void add_provider(const char *name, const struct respondd_provider_info *provider) {
+	ENTRY key = {
+		.key = (char *)provider->request,
+		.data = NULL,
+	};
+	ENTRY *entry = hsearch(key, FIND);
+	if (!entry) {
+		struct request_type *r = malloc(sizeof(*r));
+		r->providers = NULL;
+		load_cache_time(r, provider->request);
+
+		key.data = r;
+		entry = hsearch(key, ENTER);
+	}
+
+	struct request_type *r = entry->data;
+
+	struct provider_list *pentry = malloc(sizeof(*pentry));
+	pentry->name = strdup(name);
+	pentry->provider = provider->provider;
+
+	struct provider_list **pos;
+	for (pos = &r->providers; *pos; pos = &(*pos)->next) {
+		if (strcmp(pentry->name, (*pos)->name) < 0)
+			break;
+	}
+
+	pentry->next = *pos;
+	*pos = pentry;
+}
+
+static void load_providers(void) {
+	update_time();
+
+	/* Maximum number of request types, might be made configurable in the future */
+	if (!hcreate(32)) {
```

Comment:
If providers would want to use the hash tables, exactly one would be able to use the non-reentrant version of create... which should it be?

Choosing which code part of or loaded by respondd should be allowed to use hcreate, respondd would be the only sensible choice. I don't think there's a problem with this.

I might change this anyways, as memory managed by hcreate is annoyingly not seen by valgrind, so valgrind always reports lost allocations...